File: Hash-Based Searching/Cuckoo Hashing Search/CuckooHashingSearch.java
public class CuckooHashingSearch {

    private static final int TABLE_SIZE = 11;
    private static final int EMPTY = -1;
    private static final int MAX_KICKS = 20;

    private final int[] table1;
    private final int[] table2;

    public CuckooHashingSearch() {
        table1 = new int[TABLE_SIZE];
        table2 = new int[TABLE_SIZE];

        for (int i = 0; i < TABLE_SIZE; i++) {
            table1[i] = EMPTY;
            table2[i] = EMPTY;
        }
    }

    private int hash1(int key) {
        return key % TABLE_SIZE;
    }

    private int hash2(int key) {
        return (key / TABLE_SIZE) % TABLE_SIZE;
    }

    public boolean search(int key) {
        return table1[hash1(key)] == key ||
               table2[hash2(key)] == key;
    }

    public boolean insert(int key) {
        int current = key;
        int table = 1;

        for (int i = 0; i < MAX_KICKS; i++) {
            if (table == 1) {
                int pos = hash1(current);

                if (table1[pos] == EMPTY) {
                    table1[pos] = current;
                    return true;
                }

                int temp = table1[pos];
                table1[pos] = current;
                current = temp;
                table = 2;
            } else {
                int pos = hash2(current);

                if (table2[pos] == EMPTY) {
                    table2[pos] = current;
                    return true;
                }

                int temp = table2[pos];
                table2[pos] = current;
                current = temp;
                table = 1;
            }
        }

        return false;
    }

    public static void main(String[] args) {
        CuckooHashingSearch hash =
            new CuckooHashingSearch();

        hash.insert(20);
        hash.insert(50);
        hash.insert(53);

        System.out.println(
            hash.search(53)
                ? "Found"
                : "Not Found"
        );
    }
}

File: Hash-Based Searching/Cuckoo Hashing Search/CuckooHashingSearch.py
TABLE_SIZE = 11
EMPTY = -1
MAX_KICKS = 20


class CuckooHash:
    def __init__(self):
        self.table1 = [EMPTY] * TABLE_SIZE
        self.table2 = [EMPTY] * TABLE_SIZE

    def hash1(self, key):
        return key % TABLE_SIZE

    def hash2(self, key):
        return (key // TABLE_SIZE) % TABLE_SIZE

    def search(self, key):
        return (
            self.table1[self.hash1(key)] == key
            or self.table2[self.hash2(key)] == key
        )

    def insert(self, key):
        current = key
        table = 1

        for _ in range(MAX_KICKS):
            if table == 1:
                pos = self.hash1(current)

                if self.table1[pos] == EMPTY:
                    self.table1[pos] = current
                    return True

                self.table1[pos], current = (
                    current,
                    self.table1[pos],
                )
                table = 2

            else:
                pos = self.hash2(current)

                if self.table2[pos] == EMPTY:
                    self.table2[pos] = current
                    return True

                self.table2[pos], current = (
                    current,
                    self.table2[pos],
                )
                table = 1

        return False


if __name__ == "__main__":
    h = CuckooHash()

    h.insert(20)
    h.insert(50)
    h.insert(53)

    print("Found" if h.search(53) else "Not Found")
